src/main.rs

```diff
@@ -10,35 +10,52 @@ use libazureinit::distro::{Distribution, Distributions};
 use libazureinit::{
     error::Error as LibError,
     goalstate, imds, media,
-    media::Media,
+    media::{Environment, Media},
     reqwest::{header, Client},
     user,
 };
 
 const VERSION: &str = env!("CARGO_PKG_VERSION");
 
+// Mount the given device, get OVF environment data, return it.
+fn mount_parse_ovf_env(dev: String) -> Result<Environment, anyhow::Error> {
+    let mount_media =
+        Media::new(PathBuf::from(dev), PathBuf::from(media::PATH_MOUNT_POINT));
+    let mounted = mount_media
+        .mount()
+        .with_context(|| "Failed to mount media.")?;
+
+    let ovf_body = mounted.read_ovf_env_to_string()?;
+    let environment = media::parse_ovf_env(ovf_body.as_str())?;
+
+    mounted
+        .unmount()
+        .with_context(|| "Failed to remove media.")?;
+
+    Ok(environment)
+}
+
 fn get_username(imds_body: String) -> Result<String, anyhow::Error> {
     if imds::is_password_authentication_disabled(&imds_body)? {
         // password authentication is disabled
         Ok(imds::get_username(imds_body.clone())?)
     } else {
         // password authentication is enabled
-        let mount_media = Media::new(
-            PathBuf::from(media::PATH_MOUNT_DEVICE),
-            PathBuf::from(media::PATH_MOUNT_POINT),
-        );
-        let mounted = mount_media
-            .mount()
-            .with_context(|| "Failed to mount media.")?;
 
-        let ovf_body = mounted.read_ovf_env_to_string()?;
-        let environment = media::parse_ovf_env(ovf_body.as_str())?;
+        // list of CDROM devices that is available with possible filesystems.
+        let ovf_devices = media::get_mount_device()?;
```

Comment:
consider decoupling the mount/ovf logic out of get_username()

is the cdrom only going to mounted if we detect password enabled via IMDS?

Reply:
Sounds like a good idea.
I would say, let's do that in another PR, because I don't want to leave the bug fixes in other commits unmerged for so long time. Ok?

Reply:
I agree it'd be good to move it out, and also that it's okay for it to be in another PR. I have a WIP branch tidying some of the APIs in this area anyway so I could probably add a commit moving things around here.